### PR TITLE
OSDOCS-2231: Hardware offloading for SR-IOV

### PR DESCRIPTION
**Pull request by ctauchen**
[OSDOCS-2231](https://issues.redhat.com/browse/OSDOCS-2231): OVS hardware offloading for SR-IOV.

See 4.10 RN item here: https://github.com/openshift/openshift-docs/pull/42128

Applies only to `enterprise-4.10`

Preview: https://deploy-preview-41762--osdocs.netlify.app/openshift-enterprise/latest/networking/hardware_networks/configuring-hardware-offloading.html#about-hardware-offloading_configuring-hardware-offloading

### CODE REVIEW

**Comment by ctauchen** [comment]
@zshi-redhat Can you take a look? Thanks!

**Comment by ctauchen** [comment]
@zhaozhanqi Can you start a QE review please? Thanks!

File: modules/nw-sriov-hwol-about-hardware-offloading.adoc [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Module included in the following assemblies:
+//
+// * networking/configuring-hardware-offloading.adoc
+
+:_content-type: CONCEPT
+[id="about-hardware-offloading_{context}"]
+= About hardware offloading
+
+Hardware offloading is a method for processing data that removes data processing tasks from the CPU and transfers, or "offloads", them to the dedicated data processing unit (DPU) of a network interface controller.
```
Comment (zhaozhanqi) [inline modules/nw-sriov-hwol-about-hardware-offloading.adoc:9]: `"offloads",` the `,` is needed?

File: modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Module included in the following assemblies:
+//
+// * networking/configuring-hardware-offloading.adoc
+
+:_content-type: PROCEDURE
+[id="adding-network-attachment-definition-to-pods_{context}"]
+= Adding the network attachment definition to your pods
+
+After you create the machine config pool, the `SriovNetworkPoolConfig` and `SriovNetworkNode Policy` custom resources, and the network attachment definition, you can apply these configurations to your pods by adding the network attachment definition to your pod specifications.
```
Comment (zhaozhanqi) [inline modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc:9]: s/`SriovNetworkNode Policy`/`SriovNetworkNodePolicy`

File: modules/nw-sriov-hwol-creating-sriov-policy.adoc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  namespace: openshift-sriov-network-operator
+spec:
+  deviceType: netdevice <.>
+  eSwitchMode: "switchdev" <.>
+  nicSelector:
+    deviceID: "1019"
+    rootDevices:
+    - 0000:d8:00.0
+    vendor: "15b3"
+    pfNames:
```
Comment (zhaozhanqi) [inline modules/nw-sriov-hwol-creating-sriov-policy.adoc:38]: deviceID should be `1017`
rootDevices should be changed according the real worker, So I think we need add a reference
pfNames: should be also changed according to real worker.

File: modules/nw-sriov-hwol-supported-devices.adoc [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+[cols="1,2,1,1"]
+|===
+|Manufacturer |Model |Vendor ID | Device ID
+
+|Mellanox
+|MT27800 Family [ConnectX&#8209;5]
+|15b3
+|1017
+
+|Mellanox
```
Comment (zhaozhanqi) [inline modules/nw-sriov-hwol-supported-devices.adoc:21]: do we support this one `MT28880 Family` too?  I guess we did not test this one, please correct me if I'm wrong @zshi-redhat

**Comment by zhaozhanqi** [comment]
cc @yingwang-0320 please also help take review, thanks.

File: modules/nw-sriov-hwol-supported-devices.adoc (hunk 4 above)
Comment (zshi-redhat) [reply modules/nw-sriov-hwol-supported-devices.adoc:21]: I have ConnectX-5 Ex card (MT28800 Family) in my environment, which Christopher used for collecting the device information. I was told these two cards (CX-5 and CX-5 Ex) is very similar, but let me double check with FDP team to see if they verify them separately.

File: modules/nw-sriov-hwol-about-hardware-offloading.adoc [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+In the same way that a dedicated graphics card can improve graphics performance, a smartNIC can improve data processing.
+In each case, specialized hardware takes the processing workloads away from the CPU to increase performance.
+
+In {product-title], you can configure hardware offloading for bare metal nodes that have a compatible smartNIC.
+Hardware offloading is configured and enabled by the SR-IOV Network Operator.
+
+Hardware offloading is not compatible with all workloads or application types.
+Only two communication types are supported: pod-to-pod and pod-to-service communication.
+In all cases, hardware offloading takes place only when those pods and services are assigned to nodes that have a compatible smartNIC.
+If, for example, a pod on a node with hardware offloading tries to communicate with a service on a regular node, then all of the processing will take place in the kernel, and no offloading will take place.
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-about-hardware-offloading.adoc:23]: Agreed and rewritten.

File: modules/nw-sriov-hwol-configuring-machine-config-pool.adoc [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+[source,terminal]
+----
+NAME       STATUS   ROLES            AGE   VERSION
+master-0   Ready    master           2d    v1.23.3+d99c04f
+master-1   Ready    master           2d    v1.23.3+d99c04f
+master-2   Ready    master           2d    v1.23.3+d99c04f
+worker-0   Ready    worker           2d    v1.23.3+d99c04f
+worker-1   Ready    worker           2d    v1.23.3+d99c04f
+worker-2   Ready    offload,worker   47h   v1.23.3+d99c04f
+worker-3   Ready    offload,worker   47h   v1.23.3+d99c04f
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-configuring-machine-config-pool.adoc:69]: Yes, and I've uncommented this verification step. Thanks!

File: modules/nw-sriov-hwol-about-hardware-offloading.adoc (hunk 1 above)
Comment (ctauchen) [reply modules/nw-sriov-hwol-about-hardware-offloading.adoc:9]: Agreed, that's not quite right. Will have a think and take this up with peer reviewer. Thanks!

File: modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc (hunk 2 above)
Comment (ctauchen) [reply modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc:9]: Changed, thanks!

File: modules/nw-sriov-hwol-creating-sriov-policy.adoc (hunk 3 above)
Comment (ctauchen) [reply modules/nw-sriov-hwol-creating-sriov-policy.adoc:38]: @zshi-redhat Can you comment on this? The YAML in both policy docs you provided use `deviceID: 1019`, and this was tested. Did I miss something?

File: modules/nw-sriov-hwol-about-hardware-offloading.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yingwang-0320) [reply modules/nw-sriov-hwol-about-hardware-offloading.adoc]: Hi @zshi-redhat, as per our testing,  pod to nodeport service and hostnetwork service don't support hardware offload, can we describe the supporting two communication types more accurately?
1. pod to pod
2. pod to clusterIP service backed by regular pod

**Comment by yingwang-0320** [comment]
@zhaozhanqi sure. Thank you.

File: modules/nw-sriov-hwol-about-hardware-offloading.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-about-hardware-offloading.adoc]: Thanks @yingwang-0320 @zshi-redhat ! I've made the change.

**Comment by yingwang-0320** [comment]
@zshi-redhat , in the step of "Adding the network attachment definition to your pods", the doc attach interface as below. 
....
metadata:
  annotations:
    k8s.v1.cni.cncf.io/networks: net-attach-def/net-attach-def 

But I just tested "v1.multus-cni.io/default-network" like below. 
....
metadata:
  annotations:	
    v1.multus-cni.io/default-network: offload-testing/default

I tried the first one on my cluster which shows error info "Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_test-pod-11_offload-testing_c58de843-96b6-48c4-b6fd-ac1d16e6db7e_0(63a65b2a4626ac1b2eafc057ce4f7213ecbec9ee34dbd2c0951ea53921cd838a): error adding pod offload-testing_test-pod-11 to CNI network "multus-cni-network": plugin type="multus" name="multus-cni-network" failed (add): [offload-testing/test-pod-11/c58de843-96b6-48c4-b6fd-ac1d16e6db7e:ovn-kubernetes]: error adding container to network "ovn-kubernetes": CNI request failed with status 400: '[offload-testing/test-pod-11 63a65b2a4626ac1b2eafc057ce4f7213ecbec9ee34dbd2c0951ea53921cd838a] [offload-testing/test-pod-11 63a65b2a4626ac1b2eafc057ce4f7213ecbec9ee34dbd2c0951ea53921cd838a] failed to configure pod interface: failed to rename b8cd3b21bd257f3 to 63a65b2a4626ac1: file exists
".

Can you please help confirm

**Comment by zshi-redhat** [comment]
> @zshi-redhat , in the step of "Adding the network attachment definition to your pods", the doc attach interface as below. `.... metadata: annotations: k8s.v1.cni.cncf.io/networks: net-attach-def/net-attach-def `
> 
> But I just tested "v1.multus-cni.io/default-network" like below. `.... metadata: annotations: v1.multus-cni.io/default-network: offload-testing/default`
> 
> Can you please help confirm

Good catch! 
Yes, the latter `v1.multus-cni.io/default-network` shall be used.

File: networking/hardware_networks/configuring-hardware-offloading.adoc [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+include::modules/nw-sriov-hwol-configuring-machine-config-pool.adoc[leveloffset=+1]
+
+//Configuring the SR-IOV network node policy
+include::modules/nw-sriov-hwol-creating-sriov-policy.adoc[leveloffset=+1]
+
+//Creating a Network Attachment Definition
+include::modules/nw-sriov-hwol-creating-network-attachment-definition.adoc[leveloffset=+1]
+
+//Adding the network attachment definition to your pods
+include::modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc[leveloffset=+1]
```
Comment (yingwang-0320) [inline networking/hardware_networks/configuring-hardware-offloading.adoc:37]: please see comment above

File: modules/nw-sriov-hwol-configuring-machine-config-pool.adoc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+[source,terminal]
+----
+NAME       STATUS   ROLES                   AGE   VERSION
+master-0   Ready    master                  2d    v1.23.3+d99c04f
+master-1   Ready    master                  2d    v1.23.3+d99c04f
+master-2   Ready    master                  2d    v1.23.3+d99c04f
+worker-0   Ready    worker                  2d    v1.23.3+d99c04f
+worker-1   Ready    worker                  2d    v1.23.3+d99c04f
+worker-2   Ready    mcp-offloading,worker   47h   v1.23.3+d99c04f
+worker-3   Ready    mcp-offloading,worker   47h   v1.23.3+d99c04f
```
Comment (bn222) [inline modules/nw-sriov-hwol-configuring-machine-config-pool.adoc:68]: Does it make sense to have a marker <1> behind mcp-offloading here?

File: modules/nw-sriov-hwol-about-hardware-offloading.adoc [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+// * networking/configuring-hardware-offloading.adoc
+
+:_content-type: CONCEPT
+[id="about-hardware-offloading_{context}"]
+= About hardware offloading
+
+Open vSwitch hardware offloading is a method of processing data that diverts data processing tasks away from the CPU and offloads them to the dedicated data processing unit (DPU) of a network interface controller.
+As a result, clusters can benefit from faster data processing, reduced CPU workloads, and lower computing costs.
+
+The key element for this feature is a modern class of network interface controllers known as smartNICs.
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-about-hardware-offloading.adoc:12]: Thanks @bn222 . The terminology around NICs, SmartNICs, and DPUs is a bit of a pain, and there still isn't consistency among vendors. I'm thinking now that the cleanest and most future-proof method is to simply avoid naming this class of device altogether.

If I do this, I'll refer to this generally as a NIC, and more specifically, when necessary, as a NIC with data processing capabilities, or something along those lines.

How does that sit with you?

File: modules/nw-operator-cr.adoc [hunk 10]
```diff
@@ -311,6 +311,7 @@ One of the following additional audit log targets:
 |====
 // end::policy-audit[]
 
+[id="gatewayConfig-object_{context}"]
```
Comment (mjpytlak) [inline modules/nw-operator-cr.adoc:314]: There might be several instances of OpenShift that might need to be changed to `{product-title}` For example, does OpenShift SDN need to be {product-title} SDN? I did notice this is several other modules, so I was not sure if the feature name needed the variable, but at least wanted to mention it.

File: networking/hardware_networks/configuring-hardware-offloading.adoc [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+//Which NICs do we support?
+
+include::modules/nw-sriov-hwol-supported-devices.adoc[leveloffset=+1]
+
+[id="configuring-hardware-offloading-prerequisites"]
+== Prerequisites
+
+* Your cluster has at least one bare metal machine with a network interface controller that is supported for hardware offloading.
+* You xref:/installing-sriov-operator.adoc#installing-sr-iov-operator_installing-sriov-operator[installed the SR-IOV Network Operator].
```
Comment (mjpytlak) [inline networking/hardware_networks/configuring-hardware-offloading.adoc:23]: This link is broken in the published output.

File: modules/nw-sriov-hwol-configuring-machine-config-pool.adoc [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+$ oc label node <node_name> node-role.kubernetes.io/<offload_name>=""
+----
+
+. Optional: To verify that the new pool is created, run the following command:
+[source,terminal]
+----
+$ oc get nodes
+----
++
+.Example output
```
Comment (mjpytlak) [inline modules/nw-sriov-hwol-configuring-machine-config-pool.adoc:58]: The label renders as `+ .Example output` in the published output.

https://deploy-preview-41762--osdocs.netlify.app/openshift-enterprise/latest/networking/hardware_networks/configuring-hardware-offloading.html#about-hardware-offloading_configuring-hardware-offloading

File: modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+:_content-type: PROCEDURE
+[id="adding-network-attachment-definition-to-pods_{context}"]
+= Adding the network attachment definition to your pods
+
+After you create the machine config pool, the `SriovNetworkPoolConfig` and `SriovNetworkNodePolicy` custom resources, and the network attachment definition, you can apply these configurations to your pods by adding the network attachment definition to your pod specifications.
+
+.Procedure
+
+. In the pod specification, add the `.metadata.annotations.k8s.v1.cni.cncf.io/networks` field and specify the network attachment definition you created for hardware offloading:
```
Comment (mjpytlak) [inline modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc:13]:
```suggestion
* In the pod specification, add the `.metadata.annotations.k8s.v1.cni.cncf.io/networks` field and specify the network attachment definition you created for hardware offloading:
```
Single-step procedures should use an unordered list.

File: modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc [hunk 14]
```diff
@@ -0,0 +1,22 @@
+// Module included in the following assemblies:
+//
+// * networking/configuring-hardware-offloading.adoc
+
+:_content-type: PROCEDURE
+[id="adding-network-attachment-definition-to-pods_{context}"]
+= Adding the network attachment definition to your pods
+
+After you create the machine config pool, the `SriovNetworkPoolConfig` and `SriovNetworkNodePolicy` custom resources, and the network attachment definition, you can apply these configurations to your pods by adding the network attachment definition to your pod specifications.
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc:9]: Hmm. I see what you mean. What I was going for here is just to situate this final procedure in the overall process. Strictly speaking, these are all prerequisites. But if we go down that route, every procedure in this assembly needs all the previous procedures listed as prerequisites. I'm not sure whether that's necessary.

File: modules/nw-operator-cr.adoc (hunk 10 above)
Comment (ctauchen) [reply modules/nw-operator-cr.adoc:314]: This is the name of a container network interface, one of two that we provide. I believe the CNI name is the same regardless of what platform you use it on. Will double check with other networking folks.

File: modules/nw-sriov-hwol-about-hardware-offloading.adoc [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+Hardware offloading is configured and enabled by the SR-IOV Network Operator.
+
+Hardware offloading is not compatible with all workloads or application types.
+Only the following two communication types are supported:
+* pod-to-pod
+* pod-to-service, where the service is a ClusterIP service backed by a regular pod
+
+In all cases, hardware offloading takes place only when those pods and services are assigned to nodes that have a compatible smartNIC.
+Suppose, for example, that a pod on a node with hardware offloading tries to communicate with a service on a regular node.
+On the regular node, all the processing takes place in the kernel, so the overall performance of the pod-to-service communication is limited to the maximum performance of that regular node.
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-about-hardware-offloading.adoc:27]: Thanks, I'll consider whether accounting for that situation is useful for our users.

File: modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc (hunk 13 above)
Comment (ctauchen) [reply modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc:13]: Thanks for catching that! Fixed!

File: modules/nw-sriov-hwol-configuring-machine-config-pool.adoc [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+.Prerequisites
+
+* You installed the OpenShift CLI (`oc`).
+* You have access to the cluster as a user with the `cluster-admin` role.
+
+.Procedure
+
+. Create a machine config pool for machines you want to use hardware offloading on.
+
+.. Create a file, such as `mcp-offloading.yaml`, with content like the following example:
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-configuring-machine-config-pool.adoc:20]: Stet. I prefer to keep the explanation for the example YAML.

File: modules/nw-sriov-hwol-configuring-machine-config-pool.adoc [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+master-2   Ready    master                  2d    v1.23.3+d99c04f
+worker-0   Ready    worker                  2d    v1.23.3+d99c04f
+worker-1   Ready    worker                  2d    v1.23.3+d99c04f
+worker-2   Ready    mcp-offloading,worker   47h   v1.23.3+d99c04f
+worker-3   Ready    mcp-offloading,worker   47h   v1.23.3+d99c04f
+----
+
+. Add this machine config pool to the `SriovNetworkPoolConfig` custom resource:
+
+.. Create a file, such as `sriov-pool-config.yaml`, with content like the following example:
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-configuring-machine-config-pool.adoc:73]: Stet. I prefer to keep the explanation for the example YAML.

File: modules/nw-sriov-hwol-creating-network-attachment-definition.adoc [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+After you define the machine config pool and the SR-IOV network node policy, you can create a network attachment definition for the network interface card you specified.
+
+.Prerequisites
+
+* You installed the OpenShift CLI (`oc`).
+* You have access to the cluster as a user with the `cluster-admin` role.
+
+.Procedure
+
+. Create a file, such as `net-attach-def.yaml`, with content like the following example:
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-creating-network-attachment-definition.adoc:18]: Stet. I prefer to keep the explanation for the example YAML.

File: modules/nw-sriov-hwol-creating-sriov-policy.adoc [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+The following procedure creates an SR-IOV interface for a network interface controller with hardware offloading.
+
+.Prerequisites
+
+* You installed the OpenShift CLI (`oc`).
+* You have access to the cluster as a user with the `cluster-admin` role.
+
+.Procedure
+
+. Create a file, such as `sriov-node-policy.yaml`, with content like the following example:
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-hwol-creating-sriov-policy.adoc:21]: Stet. I prefer to keep the explanation for the example YAML.

File: modules/nw-sriov-networknodepolicy-object.adoc [hunk 20]
```diff
@@ -32,7 +32,8 @@ spec:
     netFilter: "<filter_string>" <14>
   deviceType: <device_type> <15>
   isRdma: false <16>
-  linkType: <link_type> <17>
+    linkType: <link_type> <17>
```
[the remark this answers is not in the document]
Comment (ctauchen) [reply modules/nw-sriov-networknodepolicy-object.adoc:35]: Done.

File: networking/hardware_networks/configuring-hardware-offloading.adoc (hunk 11 above)
Comment (ctauchen) [reply networking/hardware_networks/configuring-hardware-offloading.adoc:23]: Fixed.

File: modules/nw-sriov-hwol-configuring-machine-config-pool.adoc (hunk 12 above)
Comment (ctauchen) [reply modules/nw-sriov-hwol-configuring-machine-config-pool.adoc:60]: Fixed, thanks!

File: modules/nw-sriov-hwol-supported-devices.adoc (hunk 4 above)
Comment (ctauchen) [reply modules/nw-sriov-hwol-supported-devices.adoc:21]: @zhaozhanqi I believe both cards have been now been confirmed by QE.

Comment (zshi-redhat) [reply modules/nw-sriov-hwol-supported-devices.adoc:21]: @zhaozhanqi FDP team verifies OVS HWOL feature for both CX-5 and CX-5 Ex cards, we shall be fine to support them in OCP.

File: modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc (hunk 14 above)
[the remark this answers is not in the document]
Comment (mjpytlak) [reply modules/nw-sriov-hwol-adding-network-attachment-definitions-to-pods.adoc:9]: I do not think your Prerequisites need to cascade, if you will. For example if I have procedure A, B, C, D, completing A may be required to complete B; however (considering that modules represent content that can stand alone), completing D might not necessary rely on A and B, perhaps only C. Feel free to ignore. I do see now how you are situating the final step.

File: modules/nw-sriov-hwol-supported-devices.adoc (hunk 4 above)
Comment (zhaozhanqi) [reply modules/nw-sriov-hwol-supported-devices.adoc:21]: ok, Great,  thanks.

File: modules/nw-sriov-hwol-creating-sriov-policy.adoc (hunk 3 above)
Comment (zhaozhanqi) [reply modules/nw-sriov-hwol-creating-sriov-policy.adoc:38]: it's ok for me with 1019.

**Comment by zhaozhanqi** [comment]
> @zhaozhanqi Now that we've resolved the issue with the two cards, can you provide a final review? Thanks!

/lgtm

File: modules/nw-operator-cr.adoc (hunk 10 above)
Comment (ctauchen) [reply modules/nw-operator-cr.adoc:314]: Definitely `OpenShift SDN`. STET.

**Comment by mjpytlak** [comment]
@ctauchen peer review LGTM.

**Pull request merged**